v4: vi.mocked() regression for static class methods types

### Describe the bug

`vitest@4+` breaks the TS types of `.mockImplementation()`/`.mockResolvedValue()`/etc. for static class methods of `vi.mocked(MyClass)`. This was working in `vitest@3`
(tests are passing, but TS is not happy)

```ts
// Foo.ts
export class Foo {
  public static getBar(): string {
    return 'bar';
  }
}
```

```ts
// Foo.test.ts
import { vi, test, expect } from 'vitest';
import { Foo } from './Foo.js';

vi.mock('./Foo');

const mockedFoo = vi.mocked(Foo, { partial: true });

test('test', () => {
  mockedFoo.getBar.mockImplementation(() => 'mocked bar');

  expect(mockedFoo.getBar()).toBe('mocked bar');
});
```

```
 error TS2339: Property 'mockImplementation' does not exist on type ...
```

<img width="505" height="217" alt="Image" src="https://github.com/user-attachments/assets/43371091-c1b3-4232-9234-9301d5e5b100" />

### Reproduction

[Repro StackBlitz link](https://stackblitz.com/edit/vitest-dev-vitest-zzeqxagh?file=src%2FFoo.test.ts)

### System Info

```shell
System:
    OS: Linux 5.0 undefined
    CPU: (8) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
    Memory: 0 Bytes / 0 Bytes
    Shell: 1.0 - /bin/jsh
  Binaries:
    Node: 20.19.1 - /usr/local/bin/node
    Yarn: 1.22.19 - /usr/local/bin/yarn
    npm: 10.8.2 - /usr/local/bin/npm
    pnpm: 8.15.6 - /usr/local/bin/pnpm
  npmPackages:
    vite: ^7.2.6 => 7.2.6 
    vitest: ^4.0.15 => 4.0.15
```

### Used Package Manager

npm

### Validations

- [x] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [x] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [x] Read the [docs](https://vitest.dev/guide/).
- [x] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [x] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [x] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

--- a/packages/spy/src/types.ts
+++ b/packages/spy/src/types.ts
@@ -462,12 +462,10 @@ type DeepPartialMock<T extends Procedure | Constructable = Procedure> = Mock<
 export type MaybeMockedConstructor<T> = T extends Constructable
   ? Mock<T>
   : T
-export type MockedFunction<T extends Procedure | Constructable> = Mock<T> & {
-  [K in keyof T]: T[K];
-}
-export type PartiallyMockedFunction<T extends Procedure | Constructable> = PartialMock<T> & {
-  [K in keyof T]: T[K];
-}
+export type MockedFunction<T extends Procedure | Constructable> = Mock<T>
+  & MockedObject<T>
+export type PartiallyMockedFunction<T extends Procedure | Constructable> = PartialMock<T>
+  & MockedObject<T>
 export type MockedFunctionDeep<T extends Procedure | Constructable> = Mock<T>
   & MockedObjectDeep<T>
 export type PartiallyMockedFunctionDeep<T extends Procedure | Constructable> = DeepPartialMock<T>

--- a/test/core/test/vi.spec.ts
+++ b/test/core/test/vi.spec.ts
@@ -151,6 +151,10 @@ describe('testing vi utils', () => {
       public getBar(): string {
         return this.bar
       }
+
+      static getBaz(): string {
+        return 'baz'
+      }
     }
     class FooMock implements Mocked<Foo> {
       readonly barMock: Mock<() => string> = vi.fn()
@@ -168,6 +172,8 @@ describe('testing vi utils', () => {
     if (0) {
       vi.mocked(Foo).mockImplementation(FooMock)
       vi.mocked(Foo).mockImplementation(Foo)
+      vi.mocked(Foo).getBaz.mockImplementation(() => 'baz')
+      vi.mocked(Foo, { partial: true }).getBaz.mockImplementation(() => 'baz')
     }
   })
 

